HDDS-2988. Use getPropertiesByPrefix instead of regex in matching ratis client and server properties.

bharatviswa504 opened the pull request:
## What changes were proposed in this pull request?

1. Remove getValByRegex usage and regex usage.
2. Use getByPrefix.
3. Use prefix for Ratis Client, GRPC and Serve Conf.


## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-2988

## How was this patch tested?

UT's are modified to test the behavior.

elek commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/ratis/RatisHelper.java:
```diff
@@ -54,29 +57,34 @@
-  String RATIS_GRPC_CLIENT_HEADER_REGEX = "raft\\.grpc\\.(?!server|tls)" +
-      "([a-z\\.]+)";
+  // Prefix for Ratis Server GRPC and Ratis client conf.
+  String HDDS_DATANODE_RATIS_PREFIX_KEY = "datanode.ratis.";
```
Nit: `HDDS_DATANODE_RATIS_PREFIX_KEY` can be reused.

```
  String HDDS_DATANODE_RATIS_PREFIX_KEY = "datanode.ratis.";
  String HDDS_DATANODE_RATIS_SERVER_PREFIX_KEY =
      HDDS_DATANODE_RATIS_PREFIX_KEY + ".server";
  String HDDS_DATANODE_RATIS_CLIENT_PREFIX_KEY =
      HDDS_DATANODE_RATIS_PREFIX_KEY + ".client";
  String HDDS_DATANODE_RATIS_GRPC_PREFIX_KEY =
      HDDS_DATANODE_RATIS_PREFIX_KEY + ".grpc";
```

elek commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/ratis/RatisHelper.java:
```diff
@@ -54,29 +57,34 @@
-      ".([a-z\\.]+)";
-  String RATIS_SERVER_GRPC_HEADER_REGEX = "datanode\\.ratis\\.raft\\.grpc\\" +
-      ".([a-z\\.]+)";
+  String RATIS_SERVER_PREFIX_KEY =
```
Do we really need the `server` and `client` part here? Why don't we use `datanode.ratis.raft.client.*` and `datanode.ratis.raft.server.*` instead of `datanode.ratis.server.raft.server`? Is there any use case where we need to configure the client configs on the server side? (`datanode.ratis.server.raft.client.*`)

BTW, this is just a matter of taste but I would prefer to avoid substring. This seems to be more readable for me:

```
  String HDDS_DATANODE_RATIS_SERVER_TAG = "server";
  
String HDDS_DATANODE_RATIS_PREFIX_KEY = "datanode.ratis.";

  String HDDS_DATANODE_RATIS_SERVER_PREFIX_KEY =
      HDDS_DATANODE_RATIS_PREFIX_KEY + "." HDDS_DATANODE_RATIS_SERVER_TAG;


String RATIS_SERVER_PREFIX_KEY =  RATIS_SERVER_PREFIX_KEY +"." + RaftServerConfigKeys.PREFIX;
```

elek commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/ratis/RatisHelper.java:
```diff
@@ -218,58 +223,74 @@ static RaftClient newRaftClient(RpcType rpcType, RaftPeerId leader,
   static String removeDatanodePrefix(String key) {
-    return key.replaceFirst(DATANODE_RATIS_SERVER_CONFIG_PREFIX, "");
+
+    if (key.startsWith("server.")) {
```
Is there any specific reason to limit this to `server`, `client` and `grpc`? (BTW, can we use constants to here? It's not clear if they are ratis segments or ozone segments...)

bharatviswa504 replied:
Because first, we call getByPrefix is called with "datanode.ratis." and which it returns a map by removing 'datanode.ratis." from the key, so here we will be left with 3 prefixes server/client and grpc (As ratis currently has 3 groups one for Ratis Server, Ratis Client and other for Ratis Grpc). This is the reason for the check and then based on the prefix removing the prefix. Yes, I will use constants here.

bharatviswa504 commented:
> Thanks the patch @bharatviswa504
> 
> Overall I like the idea to use string manipulation instead of regexp but it's a little hard to read the structure of the configuration. Maybe because the duplication of the "server/grpc" words in `datanode.ratis.grpc.raft.grpc.flow.control` (As I asked in an other comment why do we need the first grpc?).

The reason for server, client and grpc prefix is to group the configurations by the group. So, that it will be easy to add these configurations in different classes(like have a class for Ratis Server, One for Ratis Client ..) instead of just having all the configuration in a single class.

elek commented:
> The reason for server, client and grpc prefix is to group the configurations by the group

Thanks the answer. I understand that grouping is useful. I don't understand why do we need double grouping. 

> Do we really need the server and client part here? Why don't we use `datanode.ratis.raft.client.*` and `datanode.ratis.raft.server.*` instead of `datanode.ratis.server.raft.server`? Is there any use case where we need to configure the client configs on the server side? (`datanode.ratis.server.raft.client.*` for example?)

If we already have a grouping on the ratis side why do we introduce an other one on the ozone side?

bharatviswa504 commented:
> > The reason for server, client and grpc prefix is to group the configurations by the group
> 
> Thanks the answer. I understand that grouping is useful. I don't understand why do we need double grouping.
> 
> > Do we really need the server and client part here? Why don't we use `datanode.ratis.raft.client.*` and `datanode.ratis.raft.server.*` instead of `datanode.ratis.server.raft.server`? Is there any use case where we need to configure the client configs on the server side? (`datanode.ratis.server.raft.client.*` for example?)
> 
> If we already have a grouping on the ratis side why do we introduce an other one on the ozone side?

The reasons for a prefix "datanode" is in future if OM also uses this similar approach, we need a way to distinguish which config is for which component. So, that is the reason for prefix. So, created 3 config classes, one for ratis server with datanode.ratis.server, similarly for others. I understand the ratis.server is duplicated again with this approach. If any other thoughts how to handle this?

elek commented:
>  we can use "datanode.raft.server" for RatisServer config and for client "datanode.raft.client" and for grpc "datanode.raft.grpc". In this way we can use directly ratis grouping. Is this is what you are suggesting here?

Exactly. Sorry if was not clear. It's a very small suggestion, and not a big deal. 

My impression was that it simplifies the whole code as you don't need to do all the magic with the middle `[server|grpc]` tag in the `datanode.ratis.server.raft.server`. You can use the configuration based on the ratis groups.

bharatviswa504 commented:
> > we can use "datanode.raft.server" for RatisServer config and for client "datanode.raft.client" and for grpc "datanode.raft.grpc". In this way we can use directly ratis grouping. Is this is what you are suggesting here?
> 
> Exactly. Sorry if was not clear. It's a very small suggestion, and not a big deal.
> 
> My impression was that it simplifies the whole code as you don't need to do all the magic with the middle `[server|grpc]` tag in the `datanode.ratis.server.raft.server`. You can use the configuration based on the ratis groups.

Thank You @elek for review. Addressed review comments.

(With this approach, not able to use direct Ratis variables. (But I think that is okay)
